### DISCUSSION TITLE
Recommended way to access Workflow server from a third party client

### DISCUSSION THREAD

**Discussion by RainerAtSpirit** [elsa-workflows/elsa-core#2902] [Q&A]
We have the need to access the Elsa workflow server from a client outside of the designer. 
One option that we tried was converting the swagger.json via `openapi-generator` and the `typescript-fetch` generator. 

```
openapi-generator-cli  generate -i   libs/workflow/openapi/swagger.json --generator-name typescript-fetch -o dist/libs/elsa-api --config elsa-api.json
```

elsa-api.json
```json
{
  "generateAliasAsModel": true,
  "modelPropertyNaming": "original",
  "npmName": "@our-private-repo/elsa-api",
  "npmVersion": "1.0.0",
  "snapshot": false,
  "supportsES6": true,
  "typescriptThreePlus": true,
  "useSingleRequestParameter": true,
  "withInterfaces": false,
  "withoutRuntimeChecks":  true
}
```

However, there are a couple of inconsistencies in using that approach. The one that's causing the most issues is that swagger defines the `VersionOptions` as an object, while the actual server API expects that to be a querystring.

![image](https://user-images.githubusercontent.com/581644/160803488-f8597fcc-3207-4a59-b89d-1e13ef9388f4.png)

The ElsaClient is using the following conversion to accomplish that.
```typescript
 workflowDefinitionsApi: {
      list: async (page?: number, pageSize?: number, versionOptions?: VersionOptions) => {
        const queryString = {
          version: getVersionOptionsString(versionOptions)
        };

        if (!!page || page === 0)
          queryString['page'] = page;

        if (!!pageSize)
          queryString['pageSize'] = pageSize;

        const queryStringItems = collection.map(queryString, (v, k) => `${k}=${v}`);
        const queryStringText = queryStringItems.length > 0 ? `?${queryStringItems.join('&')}` : '';
        const response = await httpClient.get<PagedList<WorkflowDefinitionSummary>>(`v1/workflow-definitions${queryStringText}`);

        return response.data;
      },
```

Not sure if it would be possible to keep the swagger definition in sync with what the server expects. 
As an alternative, would it be possible to export `createElsaClient` and `createHttpClient` in an extra `elsa-client-api` package?

**Comment by sfmskywalker** [comment]
The Elsa API endpoint expects the version options as a string, so what we should do is fix the swagger documentation so that it does the same. This could probably be done using one of Swashbuckle's extensibility options.

I'll create an issue for this.

**Reply by sfmskywalker** [reply]
Issue: #2912
